Most recently used tab with Ctrl-Tab cycling

guifel opened the pull request:
<!-- Add brave-browser issue bellow that this PR will resolve -->
Resolves brave/brave-browser#913

I implemented MRU cycling feature as requested in the issue. I added a new option under the section Tab "Cycle through the most recently used tabs with Ctrl-Tab" to activate the feature. Deactivated by default.

This is inspired by how MRU cycling is implemented at the ChromeOS level.  

![Screenshot from 2020-09-27 12-08-46](https://user-images.githubusercontent.com/6786187/94354858-48c94880-00ba-11eb-9964-8c6c6265668a.png)


## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

TBD

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

guifel commented:
To do for me:
- [ ] Fix and adapt this for MacOS. I don't have a Mac at the moment, I might get one next week.
- [ ] Add build flags to not apply these changes for iOS and Android
- [ ] Support regular previous/next cycling while using the page up/down keys (Ctrl-Page up/down) when MRU is activated. 
- [ ] Figure out the bug with the gtk event loop crashing the browser when spamming keys on the new page tab (Might be unrelated to this)

guifel commented:
Ufffff, that was harder than I first expected but I've now implemented everything I wanted :)

First, I am a daily Brave user and what you're doing is awesome. This was my biggest frustration when moving from Firefox so I had a try. It's my first contribution to Brave and experience with the Chromium code base. It is certainly not yet perfect, I need some feedback.

I have build and tested the feature on Linux (Ubuntu), Windows and Mac. 

The only missing piece is that Ctrl-Previous and Ctrl-Next are doing MRU switching as well. I originally wanted to do a simple switch to the previous and next tab even when the feature is activated. Unfortunately, Chromium treats Ctrl-Next and Ctrl-Previous the same way as Ctrl-Tab and Ctrl-Maj-Tab down to some quite lower level controllers. If we really need this feature we will have to do some more changes at the Chromium level.

petemill commented:
Thank you so much for getting deep in to this issue @guifel and persisting to get it working! I'll definitely review the UX impact and have a go at understanding if we can split the keyboard shortcuts as you mention.

Special thanks for also putting the effort in to do the patching and overrides in a clean way.

yrliou commented on patches/chrome-browser-ui-tabs-tab_strip_model.h.patch:
```diff
@@ -0,0 +1,23 @@
+   void CloseSelectedTabs();
+ 
+   // Select adjacent tabs
++  virtual
```
This could be avoided by using chromium_src override, example: https://github.com/brave/brave-core/blob/9e7f5047afd7fdbb25aa884005918382f381578d/chromium_src/chrome/browser/ui/views/profiles/avatar_toolbar_button.h#L26

yrliou commented on patches/chrome-browser-ui-tabs-tab_strip_model.h.patch:
```diff
@@ -0,0 +1,23 @@
++  virtual
+   void SelectNextTab(
+       UserGestureDetails detail = UserGestureDetails(GestureType::kOther));
++  virtual
```
DITTO.

yrliou commented on patches/chrome-browser-ui-browser.cc.patch:
```diff
@@ -0,0 +1,21 @@
+ #include "base/threading/thread_task_runner_handle.h"
+ #include "base/time/time.h"
+ #include "build/build_config.h"
++#include "brave/browser/ui/tabs/brave_tab_strip_model.h"
```
header include should be put in a chromium_src override.

yrliou commented on patches/chrome-browser-ui-tabs-tab_strip_model.h.patch:
```diff
@@ -0,0 +1,23 @@
+   bool ShouldResetOpenerOnActiveTabChange(content::WebContents* contents) const;
+ 
+  private:
++ friend class BraveTabStripModel;
```
To be extensible, please follow this pattern:
https://github.com/brave/brave-core/blob/master/patches/chrome-browser-content_settings-content_settings_manager_impl.h.patch
https://github.com/brave/brave-core/blob/master/chromium_src/chrome/browser/content_settings/content_settings_manager_impl.h

iefremov commented on browser/ui/tabs/brave_tab_strip_model.h:
```diff
@@ -0,0 +1,61 @@
+  int current_mru_cycling_index = -1;
+
+  // List of tab indexes sorted by most recently used
+  std::vector<int> mru_cycle_list;
```
have you checked `base::MRUCache`?

guifel replied:
I was not aware of this class but I am not sure it is a good fit here. I am calculating from scratch the ordered set of tab indices at the beginning of each cycling session with `std:sort` and the timestamp of last use.
`base:MRUCache` keeps track of the ordering by calling the `Put` function whenever the item is used.

To use `MRUCache` I would need to call `Put` each time a tab is accessed, not sure this is possible form `TabStripModel`.

keur replied:
Hmm. So it looks like that timestamp is being set in `TabStripModel::ActivateTabAt`.  What about overriding that method and calling `Put` when the new tab is activated?

guifel replied:
My problem with `base::MRUCache` is that it's a map data-structure with key-value pairs. What I need is a simple MRU queue.
Also, I thought that relying on this already existing timestamp would be better than maintaining an additional index.

iefremov commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,147 @@
+
+
+BraveTabStripModel::CtrlReleaseHandler::CtrlReleaseHandler(
+  BraveTabStripModel* tab_strip): tab_strip(tab_strip) {}
```
general advice: there is a formatting tool that plays well with Google style guide - `depot_tools/clang-format` - pls run it for all files.

guifel replied:
Done, I had only ran cpplint from the npm script of brave-browser.

guifel commented:
@yrliou @iefremov do you have anymore feedback ? Is anything blocking you ?

Helldembez commented:
@guifel You are my hero.

guifel commented:
Thanks @Helldembez  :rofl: but the PR is not yet merged. Brave team seems busy.

bsclifton commented:
@guifel do you think you can give this a rebase and squash the commits up a bit? I wouldn't squash into one commit, but try to group similar ones together (ex: where you fix review feedback). Let me know if you need help with that 😄

guifel commented:
@bsclifton done. In the first commit I add the toggle button in the settings (which touches many files..) and in the second commit we have the core logic of the MRU tab switch. The 2 last commits are the requested changes by @yrliou and @iefremov.

guifel commented:
Just rebased.
Any feedback @bsclifton @yrliou @iefremov @petemill ? Let me know if anything is blocking.

bsclifton commented:
Thanks for rebasing an keeping us in the loop, @guifel - apologies for taking a while to get a review. We're working through things as best we can- we will get to this, I promise 😄

andrewklimek commented:
I'm getting a little impatient for this one since the last 2 Vivaldi snapshots are broken on my mac.  @guifel  do you have a build I could test?

guifel commented:
I just did a rebase.

The Brave team seems busy, @bsclifton any timeline ? This feature seems quite demanded.

In the meantime @andrewklimek feel free to checkout my brave-core branch `guifel:guifel-1/mru-clycling-ctrl-tab`  and build it. I do not have a mac at the moment but I tested the PR a while back.

bsclifton commented:
Apologies that it's taken a while to get to this ☹️ A lot of times I am the person reviewing and my computer broke last month. I'm working on a temporary machine until I get a new machine (repair stalled indefinitely because of repair facility closure).  Thanks for hanging in there and rebasing @guifel - it's very much appreciated 😄

keur commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+      event->type() == ui::ET_KEY_RELEASED) {
+    // Ctrl key was released, stop the MRU cycling
+    tab_strip->StopMRUCycling();
+  } else if ((event->key_code() == ui::VKEY_TAB &&
```
we can invert this statement and stop propagation that way

bridiver commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+#include "chrome/browser/profiles/profile.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/events/event.h"
```
it looks like you're missing deps for some of this stuff in browser/ui/BUILD.gn
I would suggest creating a new BUILD.gn in browser/ui/tabs so we can keep the deps organized a little better. brave_tab_menu_model.cc/brave_tab_menu_model.h should also be included in that file along with any deps required for them. Don't forget to add conditional deps for `use_aura` and `is_mac` in gn

guifel replied:
Added a new BUILD.gn in browser/ui/tabs. I am not sure what should be added in the deps. Could you point me to some docs/guidelines of what should be there ? Currently, I added the minimum for it to compile.

bridiver commented on chromium_src/chrome/browser/ui/browser.cc:
```diff
@@ -5,6 +5,7 @@
 
 #include "brave/browser/ui/brave_browser_command_controller.h"
 #include "brave/browser/ui/brave_browser_content_setting_bubble_model_delegate.h"
+#include "brave/browser/ui/tabs/brave_tab_strip_model.h"
```
why is this added here?

bridiver replied:
I believe this is where you want to put it https://github.com/brave/brave-core/pull/4671/files#diff-3a289b88edecd27ffa752266d52b41aaR8

bridiver replied:
I think what you want to do here is to create a chromium_src override for `chrome/browser/ui/browser.cc` and put the include there. That's the standard method for adding an include without patching

bridiver replied:
oops, I looked at this wrong :)

keur commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
nit: 2020

keur commented on browser/ui/tabs/brave_tab_strip_model.h:
```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
2020

keur commented on chromium_src/chrome/browser/ui/tabs/tab_strip_model.h:
```diff
@@ -0,0 +1,17 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
```
2020

simonhong commented:
From the high level point of view, handling event in model seems strange. WDYT?

simonhong commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+
+void BraveTabStripModel::SelectNextTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
```
You can do this more simply - `profile()->GetPrefs()->GetBoolean(k...)`

simonhong commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+    window->RemovePreTargetHandler(ctrl_released_event_handler.get());
+#endif
+
+    ctrl_released_event_handler.reset();
```
I can see below crashes frequently while shutdown. Maybe there is some bug for reg/unreg event handler.
```
[16574:775:0410/123156.987706:FATAL:event_handler.cc(26)] Check failed: !check_targets_ || targets_installed_on_.empty().
0   libbase.dylib                       0x000000010f24e21f base::debug::CollectStackTrace(void**, unsigned long) + 31
1   libbase.dylib                       0x000000010ef00fcb base::debug::StackTrace::StackTrace(unsigned long) + 75
2   libbase.dylib                       0x000000010ef0104d base::debug::StackTrace::StackTrace(unsigned long) + 29
3   libbase.dylib                       0x000000010ef01028 base::debug::StackTrace::StackTrace() + 40
4   libbase.dylib                       0x000000010ef4e3d7 logging::LogMessage::~LogMessage() + 183
5   libbase.dylib                       0x000000010ef4d055 logging::LogMessage::~LogMessage() + 21
6   libevents.dylib                     0x000000017099c03a ui::EventHandler::~EventHandler() + 346
7   libchrome_dll.dylib                 0x000000011706a068 BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 24
8   libchrome_dll.dylib                 0x000000011706a085 BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 21
9   libchrome_dll.dylib                 0x000000011706a0ac BraveTabStripModel::CtrlReleaseHandler::~CtrlReleaseHandler() + 28
10  libchrome_dll.dylib                 0x000000011706bacf std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler>::operator()(BraveTabStripModel::CtrlReleaseHandler*) const + 47
11  libchrome_dll.dylib                 0x0000000117069fbd std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::reset(BraveTabStripModel::CtrlReleaseHandler*) + 109
12  libchrome_dll.dylib                 0x000000011706a309 std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::~unique_ptr() + 25
13  libchrome_dll.dylib                 0x0000000117069575 std::__1::unique_ptr<BraveTabStripModel::CtrlReleaseHandler, std::__1::default_delete<BraveTabStripModel::CtrlReleaseHandler> >::~unique_ptr() + 21
14  libchrome_dll.dylib                 0x0000000117069547 BraveTabStripModel::~BraveTabStripModel() + 71
15  libchrome_dll.dylib                 0x0000000117069595 BraveTabStripModel::~BraveTabStripModel() + 21
16  libchrome_dll.dylib                 0x00000001170695bc BraveTabStripModel::~BraveTabStripModel() + 28
17  libchrome_dll.dylib                 0x000000011cec441f std::__1::default_delete<TabStripModel>::operator()(TabStripModel*) const + 47
18  libchrome_dll.dylib                 0x000000011cec439d std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::reset(TabStripModel*) + 109
19  libchrome_dll.dylib                 0x000000011cec4329 std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::~unique_ptr() + 25
20  libchrome_dll.dylib                 0x000000011ceb5375 std::__1::unique_ptr<TabStripModel, std::__1::default_delete<TabStripModel> >::~unique_ptr() + 21
21  libchrome_dll.dylib                 0x000000011ceb49a2 Browser::~Browser() + 2002
22  libchrome_dll.dylib                 0x000000011ceb53f5 Browser::~Browser() + 21
23  libchrome_dll.dylib                 0x000000011ceb551c Browser::~Browser() + 28
24  libchrome_dll.dylib                 0x000000011d45552f std::__1::default_delete<Browser>::operator()(Browser*) const + 47
25  libchrome_dll.dylib                 0x000000011d4554ed std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::reset(Browser*) + 109
26  libchrome_dll.dylib                 0x000000011d455479 std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::~unique_ptr() + 25
27  libchrome_dll.dylib                 0x000000011d444495 std::__1::unique_ptr<Browser, std::__1::default_delete<Browser> >::~unique_ptr() + 21
28  libchrome_dll.dylib                 0x000000011d443e5c BrowserView::~BrowserView() + 1836
29  libchrome_dll.dylib                 0x000000011707b4da BraveBrowserView::~BraveBrowserView() + 42
30  libchrome_dll.dylib                 0x000000011707adbc BraveBrowserView::~BraveBrowserView() + 28
31  libchrome_dll.dylib                 0x000000011707adec BraveBrowserView::~BraveBrowserView() + 28
32  libviews.dylib                      0x0000000183a74907 views::View::~View() + 503
33  libviews.dylib                      0x0000000183acf192 views::NonClientView::~NonClientView() + 306
34  libviews.dylib                      0x0000000183acf1dc views::NonClientView::~NonClientView() + 28
35  libviews.dylib                      0x0000000183acf2ec views::NonClientView::~NonClientView() + 28
36  libviews.dylib                      0x000000018385a7af std::__1::default_delete<views::View>::operator()(views::View*) const + 47
37  libviews.dylib                      0x000000018385a72d std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::reset(views::View*) + 109
38  libviews.dylib                      0x000000018385a6b9 std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::~unique_ptr() + 25
39  libviews.dylib                      0x0000000183857235 std::__1::unique_ptr<views::View, std::__1::default_delete<views::View> >::~unique_ptr() + 21
40  libviews.dylib                      0x0000000183a76a36 views::View::DoRemoveChildView(views::View*, bool, bool, bool, views::View*) + 1638
41  libviews.dylib                      0x0000000183a76ad0 views::View::RemoveAllChildViews(bool) + 112
42  libviews.dylib                      0x0000000183aac744 views::Widget::DestroyRootView() + 100
43  libviews.dylib                      0x0000000183aac4cd views::Widget::~Widget() + 109
44  libchrome_dll.dylib                 0x000000011d434e45 BrowserFrame::~BrowserFrame() + 213
45  libchrome_dll.dylib                 0x000000011708e4e8 BraveBrowserFrame::~BraveBrowserFrame() + 24
46  libchrome_dll.dylib                 0x000000011708e355 BraveBrowserFrame::~BraveBrowserFrame() + 21
47  libchrome_dll.dylib                 0x000000011708e37c BraveBrowserFrame::~BraveBrowserFrame() + 28
48  libviews.dylib                      0x0000000183b131e1 views::NativeWidgetMac::~NativeWidgetMac() + 129
49  libchrome_dll.dylib                 0x000000011d2eae3f BrowserFrameMac::~BrowserFrameMac() + 159
50  libchrome_dll.dylib                 0x000000011d2eae8c BrowserFrameMac::~BrowserFrameMac() + 28
51  libchrome_dll.dylib                 0x000000011d2eaf1c BrowserFrameMac::~BrowserFrameMac() + 28
52  libviews.dylib                      0x0000000183b1363d views::NativeWidgetMac::WindowDestroyed() + 381
53  libviews.dylib                      0x0000000183ae176c views::NativeWidgetMacNSWindowHost::OnWindowHasClosed() + 268
54  libapp_shim.dylib                   0x000000018b1d9b85 remote_cocoa::NativeWidgetNSWindowBridge::OnWindowWillClose() + 1477
55  libapp_shim.dylib                   0x000000018b1f87a0 -[ViewsNSWindowDelegate windowWillClose:] + 720
56  CoreFoundation                      0x00007fff38c6d35f __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
57  CoreFoundation                      0x00007fff38c6d2f3 ___CFXRegistrationPost1_block_invoke + 63
58  CoreFoundation                      0x00007fff38c6d268 _CFXRegistrationPost1 + 372
59  CoreFoundation                      0x00007fff38c6cebe ___CFXNotificationPost_block_invoke + 97
60  CoreFoundation                      0x00007fff38c3c7e2 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 1575
61  CoreFoundation                      0x00007fff38c3bc82 _CFXNotificationPost + 1351
62  Foundation                          0x00007fff3b2c1a22 -[NSNotificationCenter postNotificationName:object:userInfo:] + 59
63  AppKit                              0x00007fff366d4ec5 -[NSWindow _finishClosingWindow] + 185
64  AppKit                              0x00007fff361b20cd -[NSWindow _close] + 352
65  libapp_shim.dylib                   0x000000018b1d7ce7 ___ZN12remote_cocoa26NativeWidgetNSWindowBridge11CloseWindowEv_block_invoke.203 + 39
66  libapp_shim.dylib                   0x000000018b1e7a78 void base::internal::FunctorTraits<base::mac::ScopedBlock<void () block_pointer>, void>::Invoke<base::mac::ScopedBlock<void () block_pointer> >(base::mac::ScopedBlock<void () block_pointer>&&) + 88
67  libapp_shim.dylib                   0x000000018b1e7a0d void base::internal::InvokeHelper<false, void>::MakeItSo<base::mac::ScopedBlock<void () block_pointer> >(base::mac::ScopedBlock<void () block_pointer>&&) + 29
68  libapp_shim.dylib                   0x000000018b1e79c1 void base::internal::Invoker<base::internal::BindState<base::mac::ScopedBlock<void () block_pointer> >, void ()>::RunImpl<base::mac::ScopedBlock<void () block_pointer>, std::__1::tuple<> >(base::mac::ScopedBlock<void () block_pointer>&&, std::__1::tuple<>&&, std::__1::integer_sequence<unsigned long>) + 33
69  libapp_shim.dylib                   0x000000018b1e7997 base::internal::Invoker<base::internal::BindState<base::mac::ScopedBlock<void () block_pointer> >, void ()>::RunOnce(base::internal::BindStateBase*) + 87
70  libbase.dylib                       0x000000010eeaf385 base::OnceCallback<void ()>::Run() && + 117
71  libbase.dylib                       0x000000010f0ab59b base::TaskAnnotator::RunTask(char const*, base::PendingTask*) + 1563
72  libbase.dylib                       0x000000010f0fcfab base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::sequence_manager::LazyNow*, bool*) + 1627
73  libbase.dylib                       0x000000010f0fc6cf base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoSomeWork() + 159
74  libbase.dylib                       0x000000010f0fd32a non-virtual thunk to base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoSomeWork() + 42
75  libbase.dylib                       0x000000010f2a7600 base::MessagePumpCFRunLoopBase::RunWork() + 160
76  libbase.dylib                       0x000000010f2a773c ___ZN4base24MessagePumpCFRunLoopBase13RunWorkSourceEPv_block_invoke + 28
77  libbase.dylib                       0x000000010f28994a base::mac::CallWithEHFrame(void () block_pointer) + 10
78  libbase.dylib                       0x000000010f2a6583 base::MessagePumpCFRunLoopBase::RunWorkSource(void*) + 115
79  CoreFoundation                      0x00007fff38c77b21 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
80  CoreFoundation                      0x00007fff38c77ac0 __CFRunLoopDoSource0 + 103
81  CoreFoundation                      0x00007fff38c778d4 __CFRunLoopDoSources0 + 209
82  CoreFoundation                      0x00007fff38c76740 __CFRunLoopRun + 1272
83  CoreFoundation                      0x00007fff38c75bd3 CFRunLoopRunSpecific + 499
84  HIToolbox                           0x00007fff377cb65d RunCurrentEventLoopInMode + 292
85  HIToolbox                           0x00007fff377cb39d ReceiveNextEventCommon + 600
86  HIToolbox                           0x00007fff377cb127 _BlockUntilNextEventMatchingListInModeWithFilter + 64
87  AppKit                              0x00007fff35e3bba4 _DPSNextEvent + 990
88  AppKit                              0x00007fff35e3a380 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1352
89  libchrome_dll.dylib                 0x0000000118a31e8a __71-[BrowserCrApplication nextEventMatchingMask:untilDate:inMode:dequeue:]_block_invoke + 106
90  libbase.dylib                       0x000000010f28994a base::mac::CallWithEHFrame(void () block_pointer) + 10
91  libchrome_dll.dylib                 0x0000000118a31d26 -[BrowserCrApplication nextEventMatchingMask:untilDate:inMode:dequeue:] + 262
92  AppKit                              0x00007fff35e2c09e -[NSApplication run] + 658
93  libbase.dylib                       0x000000010f2a8bd8 base::MessagePumpNSApplication::DoRun(base::MessagePump::Delegate*) + 312
94  libbase.dylib                       0x000000010f2a5c15 base::MessagePumpCFRunLoopBase::Run(base::MessagePump::Delegate*) + 197
95  libbase.dylib                       0x000000010f0fe1ce base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool, base::TimeDelta) + 830
96  libbase.dylib                       0x000000010f026c8a base::RunLoop::Run() + 858
97  libchrome_dll.dylib                 0x0000000118a5b7a2 ChromeBrowserMainParts::MainMessageLoopRun(int*) + 306
98  libcontent.dylib                    0x00000001435affd9 content::BrowserMainLoop::RunMainMessageLoopParts() + 393
99  libcontent.dylib                    0x00000001435beb0a content::BrowserMainRunnerImpl::Run() + 346
100 libcontent.dylib                    0x00000001435a9b23 content::BrowserMain(content::MainFunctionParams const&) + 211
101 libcontent.dylib                    0x0000000145d61bc2 content::RunBrowserProcessMain(content::MainFunctionParams const&, content::ContentMainDelegate*) + 130
102 libcontent.dylib                    0x0000000145d635bc content::ContentMainRunnerImpl::RunServiceManager(content::MainFunctionParams&, bool) + 1532
103 libcontent.dylib                    0x0000000145d62ecb content::ContentMainRunnerImpl::Run(bool) + 731
104 libcontent.dylib                    0x0000000145d5dee8 content::ContentServiceManagerMainDelegate::RunEmbedderProcess() + 72
105 libembedder.dylib                   0x000000010ea2249f service_manager::Main(service_manager::MainParams const&) + 1903
106 libcontent.dylib                    0x0000000145d61af9 content::ContentMain(content::ContentMainParams const&) + 153
107 libchrome_dll.dylib                 0x0000000114dc5034 ChromeMain + 404
108 Brave Browser Development           0x000000010a136387 main + 583
109 libdyld.dylib                       0x00007fff7031b7fd start + 1
110 ???                                 0x0000000000000004 0x0 + 4
Task trace:
0   libapp_shim.dylib                   0x000000018b1d7833 remote_cocoa::NativeWidgetNSWindowBridge::CloseWindow() + 659
1   libchrome_dll.dylib                 0x000000011ceb904f Browser::TabStripEmpty() + 79

[0410/123157.652279:WARNING:system_snapshot_mac.cc(42)] sysctlbyname kern.nx: No such file or directory (2)
```

guifel replied:
I could reproduce this by closing the browser while cycling. This is because the `CtrlReleaseHandler` is still registered when shutting down. I solved it by calling a function that deactivate the event handler destructor check. I tried doing a standard unreg but the widget is not accessible anymore when destructing.

simonhong commented on browser/ui/tabs/brave_tab_strip_model.h:
```diff
@@ -0,0 +1,60 @@
+  std::unique_ptr<CtrlReleaseHandler> ctrl_released_event_handler;
+
+  // Current index when we are MRU cycling, set to -1 when not cycling
+  int current_mru_cycling_index = -1;
```
member variable name should be finished with `_` - `current_mru_cuycling_index_`.

guifel commented:
Thank you all for your feedback, I will address it over the weekend or early next week. 

@simonhong you are right. I spent some time today going through a ton of chromium classes. My  `CtrlReleaseHandler` class should go into a controller. Either the `BrowserTabStripController` or a new dedicated MRU controller. The calls from the view API will be made from there.

guifel commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,146 @@
+
+  if (backward) {
+    current_mru_cycling_index =
+        (current_mru_cycling_index - 1 % tabCount + tabCount) % tabCount;
```
It doesn't work because in C++ (and JS btw) `%` is the reminder operator. So `-1 % 5 = -1`.  In Python for example, you actually get the real modulo operator `-1 % 5 = 4` confusing stuff when you switch language all the time x)

guifel commented:
I addressed all your feedback, check my new commits and my comments. Sorry for the delay, I'm busy at work these days :)

simonhong commented:
@guifel Thanks for updating. I'll review tomorrow! :)

simonhong commented on chromium_src/chrome/browser/ui/tabs/tab_strip_model_observer.h:
```diff
@@ -0,0 +1,18 @@
+
+// Add virtual method to allow tab strip controllers to react upon MRU cycling
+#define BRAVE_TAB_STRIP_MODEL_OBSERVER_H_ \
+  virtual void StartMRUCycling(BraveTabStripModel* brave_tab_strip_model) {}
```
I think TabStripModelObserver is not good place to add this method.
The reason is TabStripModel has multiple TabStripModelObserver.
I think we only need one instance that implement this per TabStripModel.
Instead of expanding `TabStripModelObserver`, how about introducing new interface?

guifel replied:
I chose to use `TabStripModelObserver` because I thought that whenever `TabStripModel` needs to dispatch an action it does it through an observer. Instead, I could create a `MRUCyclingController` and move all the key handling logic there.

guifel replied:
I implemented that, please let me know what you think.

guifel commented:
@simonhong are you asking if we really need such event handlers to achieve MRU cycling ? The `BrowserCommandController` dispatching the ctrl-tab can't capture key releases. I took inspiration of how Alt-tab is implemented in Ash for ChromeOS. They use an `EventHandler` called `AltReleaseController` similar to mine: https://chromium.googlesource.com/chromium/src/+/refs/heads/master/ash/wm/window_cycle_event_filter.h#27

bridiver commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,96 @@
+
+void BraveTabStripModel::SelectNextTab(UserGestureDetails detail) {
+  bool isMRUEnabled =
+      Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())
```
please use `profile()` instead of `Profile::FromBrowserContext(GetActiveWebContents()->GetBrowserContext())`

bridiver commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,96 @@
+}
+
+void BraveTabStripModel::SelectPreviousTab(UserGestureDetails detail) {
+  bool isMRUEnabled = profile()->GetPrefs()->GetBoolean(kMRUCyclingEnabled);
```
seems like we could remove the duplication here by overriding `SelectRelativeTab` instead?

guifel replied:
Good point, I'll do that.

simonhong commented:
@guifel I tried to implement w/o event handler and pushed my testing code - https://github.com/brave/brave-core/commit/94ee9b0f5f4dfb5b5570f9c4ece4c0fbee8986e8
Can you check my code. I think it works.

guifel commented:
@simonhong I checked out and compiled your branch. The issue with this approach is that it does not differentiates the 2 following key sequences that are expected to yield different results:
- `Ctrl down, tab, tab, Ctrl up` Expected: Switch to the second most recently used tab
- `Ctrl down, Tab, Ctrl up, Ctrl down, Tab, Ctrl up`. Expected: Back to the original tab

In your solution both sequences end up on the second most recently used tab. I also tried a similar approach at first, see https://github.com/brave/brave-browser/issues/913#issuecomment-554637272
We need to capture the releases of the `Ctrl` key to differentiate 2 cycling sessions. Let me know if you are aware of a better way than using an `EventHandler` for that.

I will still reuse some parts of your code that makes it more readable than mine.

simonhong commented:
@guifel Got it. I think we can't differentiates both w/o key up event. Thanks for explanation! :)

guifel commented:
@simonhong I incorporated some of your changes, rebased and updated the patches.

They are 2 versions:
- The current one with the new `MRUTabCyclingController` 
- The version at commit https://github.com/brave/brave-core/pull/4671/commits/be74be052f94e52cfaaa27fd567bd51ca2e9cc4e with the `TabStripModelObserver`  expanded with an additional patch 

Let me know what you prefer.

simonhong commented:
@guifel  With `MRUTabCyclingController`, we have layer violation issue. `Views` API is still used in that controller.

I wanted to have dedicated one event handler instance per browser window in views layer because `TabStripModelObserver`'s purpose is for notifying model changing but MRU cycling is not related with model changing.

Instead of adding new API to `TabStripModelObserver`, how about adding to `BrowserWindow` interface? We instantiate `BrowserWindow` subclass per Browser(and TabStripModel). and I think `BraveBrowserView` is proper place to have `EventHandler`. WDYT?
`BrowsrWindow` is used to notify/change state to Views layer from outside or getting state of views layer.

If you think extending `BrowserWindow` seems reasonable you can expanding it w/o patching like below by overriding `browser_view.h`

```
/* Copyright (c) 2020 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#ifndef BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_VIEWS_FRAME_BROWSER_VIEW_H_
#define BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_VIEWS_FRAME_BROWSER_VIEW_H_

#include "chrome/browser/ui/browser_window.h"

// TODO: This abstract class can be implemented in the separate file.
class BraveBrowserWindow : public BrowserWindow {
 public:
  ~BraveBrowserWindow() override {}

  // virtual void StartMRUCycling() = 0;
};

#define BrowserWindow BraveBrowserWindow

#include "../../../../../../../chrome/browser/ui/views/frame/browser_view.h"

#undef BrowserWindow

#endif  // BRAVE_CHROMIUM_SRC_CHROME_BROWSER_UI_VIEWS_FRAME_BROWSER_VIEW_H_

```

guifel commented:
@simonhong I see. I have a problem with this.
1. How do I call `StartMRUCycling` in `BrowserWindow` from `BraveTabStripModel` ? It's possible to call functions in `TabStripObserver` because the `TabStripModel` has a list of `TabStripObserver`. But it has no reference to `BrowserWindow`
2. How to call `StopMRUCycling` of the model from the `EventHandler` in `BraveBrowserView` ? Only `BrowserTabStipController` has a reference to `TabStripModel`

simonhong commented:
@guifel We can easily fix your two issues :)

1) Usually, We can get `BrowserWindow` pointer from `Browser::window()`. By casting to `BraveBrowserWindow*`, we can call `StartMRUCycling`. To get Browser object, usually we use `BrowserFinder`. I think `BrowserFinder::FindBrowserWithWebContents()` is appropriate for now because we can know active WebContents by using `TabStripModel`.

2) You can use `browser()->tab_strip_model()` in `BraveBrowserView`.

Let me know if you have any questions. Many thanks for follow up this issue!

simonhong commented:
@rebron Can you check the UX of  new settings option? I attached captured image in the issue description.

guifel commented:
Thanks @simonhong for your help, you're the boss :) I implemented it according to your recommendations. 
Please check my latest commit.

rebron commented:
@simonhong Thanks for checking in on the pref UI. Lets not add a whole new Tab section for this one pref.
 
Good interim place for this could be after Pressing Tab in Appearance section:
<img width="696" alt="Screen Shot 2020-05-07 at 5 15 08 PM" src="https://user-images.githubusercontent.com/5951041/81356596-64d63a00-9086-11ea-9d26-ad6100d7f996.png">

simonhong commented:
@guifel Please check above @rebron 's comment :)

guifel commented:
@simonhong sorry for the delay. I added a simple browser test. I struggle testing more complex scenarios because I did not manage to emulate Ctrl Press/Release independently.  Let me know what you think of the current test.

@rebron The `Pressing Tab` option you are referring is only available on MacOS. I added the MRU option after `Always show bookmark on new tab page`. It was simpler, adding it at the bottom of the appearance section would require a new patch and more code. Let me know what you think (I'm on Linux). 
![Shot_10-Mai-2020](https://user-images.githubusercontent.com/15947729/81829334-153ab700-953b-11ea-9b27-a01a918ae319.png)

simonhong commented:
Did some minor cleanup and did run this in CI by https://github.com/brave/brave-core/pull/5546.

simonhong commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -20,6 +26,66 @@
+#endif
+
+      model_->StopMRUCycling();
+    } else if (!((event->key_code() == ui::VKEY_TAB &&
```
Still curious why other keys should be blocked?

guifel replied:
Because I assume that nothing else can be done while cycling (closing tab with shortcut, reordering tabs, etc..)
Also, Alt-tab implementation for ChromeOS does the same: https://chromium.googlesource.com/chromium/src/+/refs/heads/master/ash/wm/window_cycle_event_filter.cc#39

Btw, I just notice that we can still close tabs with the mouse while cycling. This leads to crashes, let block all clicks on the window.

simonhong replied:
I see. How about stop mru cycling when browser gets mouse event ?

Helldembez replied:
+1 for using firefox behaviour for reference.

guifel replied:
Good point!

rebron commented:
> @simonhong sorry for the delay. I added a simple browser test. I struggle testing more complex scenarios because I did not manage to emulate Ctrl Press/Release independently. Let me know what you think of the current test.
> 
> @rebron The `Pressing Tab` option you are referring is only available on MacOS. I added the MRU option after `Always show bookmark on new tab page`. It was simpler, adding it at the bottom of the appearance section would require a new patch and more code. Let me know what you think (I'm on Linux).
> ![Shot_10-Mai-2020](https://user-images.githubusercontent.com/15947729/81829334-153ab700-953b-11ea-9b27-a01a918ae319.png)

@guifel @simonhong Can we move the setting to after Page Zoom instead, which shouldn't be the very bottom. I don't like how's it's splitting the two Bookmark settings, doesn't make sense in between there.

guifel commented:
Thanks for the feedback. 
I made some testing on MacOS and I figured out a new issue. When clicking anywhere while cycling it opens the contextual menu (right-click menu) and it won't capture the Ctrl release. The event handler can't capture the click and therefore I can't block it or stop the cycling.

bridiver commented:
sorry this has been languishing here, but there are some changes in c83 that I think can help with the patches https://github.com/brave/brave-core/pull/5102/commits/b37a265f197e1b64fc5dfad459744896095ced2a

simonhong commented:
@guifel I tried to use `EventMonitor` to monitoring events - https://github.com/brave/brave-core/commit/38643cd048269f166336d5a33937092c94a51a8b WDYT? This works well on macOS. and I removed other key event monitoring code because when mouse click event happens cycling is stoped. also I used new `MRUCyclingEventHandler` class name because it also monitors window activation state. I think cycling should be stoped when windows goes inactive.
WDYT?

guifel commented:
@simonhong thanks, your approach with the event monitor works very well! I've added your commit made some minor changes. I've tested all the edge cases I'm aware of on MacOS and it looks good for now. I'm now building it on Linux and Windows will follow. 

@rebron I've moved the setting after the Page Zoom setting, this is on MacOS:
<img width="1035" alt="Settings 2020-05-28 18-25-44" src="https://user-images.githubusercontent.com/15947729/83167686-b8312a80-a110-11ea-9671-7b80d7f99d39.png">

Kwadz commented:
Any update on this? Is it blocked because of the conflicting files?

simonhong commented:
@Kwadz OMG. Completely forgot about this. I'll follow soon.

chriscarollo commented:
Any news?  MRU tabbing is literally the only reason I'm sticking with Firefox and not switching to Brave.

guifel commented:
Unfortunately I don't have the time to polish the PR for merge. But I have been using it for a while and it works well.

chriscarollo commented:
What more work needs to be done?  Just resolving the conflict?

Marr11317 commented:
Just a little curious question: Is this PR deeply implemented into the Brave architecture?
Would porting this PR to chromium (or another chromium based browser) be a lot of work?

bsclifton commented:
@Marr11317 good question - we could look at "upstreaming" this to Chromium and seeing if they accept it. They may not - but it might be worth a try

rjshrjndrn commented:
@guifel @bsclifton do you guys have any mechanism to build this pr for Arch Linux? Sorry to bother, but this PR is one of the most requested features, but seems like will get some time to get merged. I know you guys will be busy. That's why trying to build myself.

simonhong commented:
Sorry for the delay! It's ready to review again :)

rjshrjndrn commented:
wow @simonhong Thank you for picking this up. Can't wait to do `ctrl + tab` in Brave!!!

bsclifton commented:
Great job with this @guifel and @simonhong 😄 Works really nicely! Will try to review code tomorrow

bsclifton reviewed (approved):
Code looks great! 😄👍 

cc: @bridiver on patch review

bsclifton commented:
@keur can you re-review? Also curious if you can check out @iefremov 😄

keur commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,74 @@
+  if (mru_cycle_list_.empty()) {
+    // Start cycling
+
+    Browser* browser = chrome::FindBrowserWithWebContents(GetWebContentsAt(0));
```
nit: Maybe `chrome::FindLastActive()` more straightforward here?

keur commented on browser/ui/views/frame/brave_browser_view.cc:
```diff
@@ -20,6 +24,85 @@
+  // ui::EventObserver overrides:
+  void OnEvent(const ui::Event& event) override {
+    if (event.type() == ui::ET_KEY_RELEASED &&
+        (event.IsKeyEvent() &&
```
`IsKeyEvent()` is implied by `ET_KEY_RELEASED`

keur commented on browser/ui/tabs/brave_tab_strip_model.cc:
```diff
@@ -0,0 +1,74 @@
+              });
+
+    // Tell the cycling controller that we start cycling to handle tabs keys
+    static_cast<BraveBrowserWindow*>(browser->window())->StartMRUCycling();
```
this seems poorly named. It's not really starting MRUCycling. That's done here. It's listening for MRU stop events.

keur reviewed (commented):
Mostly LGTM. The code has become a lot simpler. Looking very good.

bsclifton commented:
@keur can you please re-review? 😄

simonhong commented:
Kindly ping @bridiver :)

bridiver commented on browser/ui/BUILD.gn:
```diff
@@ -366,7 +366,10 @@ source_set("ui") {
   if (toolkit_views) {
-    deps += [ "//ui/views" ]
+    deps += [
+      "//ui/events",
```
what was this added for?

bridiver commented on test/BUILD.gn:
```diff
@@ -74,6 +74,15 @@ group("other_unit_tests") {
 }
 
+if (!is_android) {
+group("other_browser_tests") {
```
`other_browser_tests` is not needed, you can add `"//brave/browser/ui/tabs/test:browser_tests"` directly as a dep of "browser_tests"

simonhong replied:
Fixed.

bridiver reviewed (approved):
patches look ok so I'm approving those as code owner, but there are some other small issues and also please address any other comments

simonhong commented:
@guifel, Finally merged!  🎉

guifel commented:
Awesome to see this feature in the master! Thank you guys for the amazing job.

bsclifton commented:
@guifel thanks for your contribution and also your patience 😄 This is a great feature!
@simonhong thanks for championing this (rebasing, fixing items, etc) and taking it across the finish line

This should be in Nightly in the next day or so 😄

rjshrjndrn commented:
woohoo at last... Here comes Brave With C-Tab

motevets commented:
FYI: I have submitted an issue to Chromium to consider merging/cherry-picking these commits into Chromium. https://bugs.chromium.org/p/chromium/issues/detail?id=1150395

bsclifton commented:
@motevets awesome, thanks! That would be a great one to upstream

believerd commented:
Hi developer, there seems to be a bug in the MRU implementation. If you open **just one** new tab using `open link in new tab`(right click) then Ctrl-Tab won't work.
